### ISSUE TITLE
JSON decode on output

### ISSUE THREAD

**Issue by konsumer** [MrcSnm/REST-love#2]
Since we have json encode on input, it might also be help for output, even via a separate function, although I would argue that REST is generally JSON, so parsing by default would be cool. Would it make sense to use a complete JSON lib, like dkjson?

**Comment by MrcSnm** [comment]
I believe it is a nice idea, but I really don't wish to use a JSON lib for that, specially that Lua's table has many things in common with Js objects, I believe that should be possible to do in a few tens of lines

**Comment by konsumer** [comment]
I use [dkjson](https://github.com/LuaDist/dkjson/blob/master/dkjson.lua) as an example because it is fairly simple, but complete & well-tested (or at least better than a few others.) An inline home-rolled json codec is fine too, I guess, my main point is that it seems weird to me to take a table (internally converted to json) as input, and output a string. Using a shared json lib also means it can be reused (like the other parts of this lib, for example, I use your module-loader in other things, it's very useful) so your users can encode/decode json in other places, in a fairly standard way. It's trivial for me to wrap every callback with `json.decode` from dkjson, so I can just do it there, it just seems like a surprising thing for users, to point out: "table in, string out". It's also not reusing the code for parsing/encoding as much as it should, so there are basically 2 implementations of json-encode, with no overlap between encode/decode. Since you already also do rudimentary JSON-encoding, it makes sense to me to use a lib, unless there is some big problem with dkjson.

I dunno, I am thinking about writing a request library that uses libcurl directly in a love-thread with FFI, so I'm not sure I will end up using REST-love, anyway. I need binary requests and some non-json text requests, and I'd rather share more logic between them, and make something that generally works the way I want.

**Comment by MrcSnm** [comment]
Humm I think the string out is great, because it would create a situation that I would need to check if the output is really a json, if it is not, then there would be no output, which is a bug.
That could be solved by having an argument which receive a boolean to encode the string into a JSON.

I believe it would be great having a direct libcurl. I didn't do that at the time because I avoided doing complicated things at the project.

About having 2 implementations, this really worries me. That need to be took care of, please point out where you're finding (I haven't been programming Lua for a bit of time already)


So basically:

I believe that I could have versions for sending JSON or binary data. For both input and output

**Comment by konsumer** [comment]
> Humm I think the string out is great, because it would create a situation that I would need to check if the output is really a json, if it is not, then there would be no output, which is a bug.

My point is REST APIs generally do JSON in/out, and you are already assuming that for input, but not output. If it doesn't return anything, than that seems like a normal case for "something went wrong".

In lua, a common error pattern is the kind of thing `pcall` outputs:

```
data, error = myfunc()
```

similar could be done in callback.

> That could be solved by having an argument which receive a boolean to encode the string into a JSON.

That seems like a good idea. I also don't mind doing my own encoding/decoding in another library, it just surprises me that they don't match.

> I believe it would be great having a direct libcurl. I didn't do that at the time because I avoided doing complicated things at the project.

I will make it and shout out to your project, in the README. I think I will take an even more minimal approach, and not do any json or anything, just so the user can do that however they want.

> About having 2 implementations, this really worries me. That need to be took care of, please point out where you're finding (I haven't been programming Lua for a bit of time already)

This is more a general programming thing, than anything specific to lua. I mean for end-users. Here is an example:

I use your REST library, it has [it's own simple JSON encoder](https://github.com/MrcSnm/REST-love/blob/master/REST.lua#L26-L56) Since I also want to decode JSON, I import a library that does that (we'll just say dkjson, for example) and it already has `json.encode` so that is 1 bit of functional duplication, but if you look at it, it reuses a bunch of code between encoder/decoder, which means that by using your library I am duplicating more implementations of things (the initial encoder and also not reusing shared functions for encode/decode.)

**Comment by konsumer** [comment]
Since you seem resistant to this idea, and I am not even going to have automatic json parsing in/out on my own library, that I'm actually going to use, I will close this issue. Obviously feel free to re-open it, if you want this enhancement.
